pkg/kubelet/cm/dra/claiminfo.go

```diff
@@ -21,78 +21,115 @@ import (
 	"sync"
 
 	"k8s.io/apimachinery/pkg/types"
-	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/kubelet/cm/dra/state"
 	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
 )
 
 // claimInfo holds information required
 // to prepare and unprepare a resource claim.
 type claimInfo struct {
-	sync.RWMutex
-
-	// name of the DRA driver
-	driverName string
-
-	// claimUID is an UID of the resource claim
-	claimUID types.UID
-
-	// claimName is a name of the resource claim
-	claimName string
-
-	// namespace is a claim namespace
-	namespace string
-
-	// podUIDs is a set of pod UIDs that reference a resource
-	podUIDs sets.Set[string]
-
-	// cdiDevices is a list of CDI devices returned by the
-	// GRPC API call NodePrepareResource
-	cdiDevices []string
-
+	state.ClaimInfoState
 	// annotations is a list of container annotations associated with
 	// a prepared resource
 	annotations []kubecontainer.Annotation
 }
 
-func (res *claimInfo) addPodReference(podUID types.UID) {
-	res.Lock()
-	defer res.Unlock()
-
-	res.podUIDs.Insert(string(podUID))
+// claimInfoCache is a cache of processed resource claims keyed by namespace + claim name.
+type claimInfoCache struct {
+	sync.RWMutex
+	state     state.CheckpointState
+	claimInfo map[string]*claimInfo
 }
 
-func (res *claimInfo) deletePodReference(podUID types.UID) {
-	res.Lock()
-	defer res.Unlock()
+func stateToCache(curState state.ClaimInfoStateList) (map[string]*claimInfo, error) {
+	cache := make(map[string]*claimInfo)
+	for _, entry := range curState {
 
-	res.podUIDs.Delete(string(podUID))
+		// NOTE: Passing CDI device names as annotations is a temporary solution
+		// It will be removed after all runtimes are updated
+		// to get CDI device names from the ContainerConfig.CDIDevices field
+		annotations, err := generateCDIAnnotations(entry.ClaimUID, entry.DriverName, entry.CdiDevices)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate container annotations, err: %+v", err)
+		}
+		res := claimInfo{ClaimInfoState: state.ClaimInfoState{DriverName: entry.DriverName,
+			ClaimUID:   entry.ClaimUID,
+			ClaimName:  entry.ClaimName,
+			Namespace:  entry.Namespace,
+			PodUIDs:    entry.PodUIDs,
+			CdiDevices: entry.CdiDevices,
+		},
+			annotations: annotations,
+		}
+		key := entry.ClaimName + entry.Namespace
+		cache[key] = &res
+	}
+	return cache, nil
 }
 
-// claimInfoCache is a cache of processed resource claims keyed by namespace + claim name.
-type claimInfoCache struct {
-	sync.RWMutex
-	claimInfo map[string]*claimInfo
+func cacheToState(cache map[string]*claimInfo) state.ClaimInfoStateList {
```

Comment:
Can we just change this function to:
```
func (cache *claimInfoCache) SyncToCheckpoint() {
...
}
```
And add this at the bottom:
```
cache.state.Store(claimInfoStateList)
```

As far as I can tell, there is never a case where we need  `cacheToState` without a call to `cache.state.Store()` immediately following it, and this change cleans up the API a bit.

Reply:
done